plugins/inputs/sqlserver/sqlserver.go

```diff
@@ -423,6 +464,96 @@ func (s *SQLServer) Init() error {
 	return nil
 }
 
+// Get Token Provider by loading cached token or refreshed token
+func getTokenProvider() (func() (string, error), error) {
+	var tokenString string
+
+	// load token
+	muCacheLock.RLock()
+	token, err := loadToken()
+	muCacheLock.RUnlock()
+
+	// if there's error while loading token or found an expired token, refresh token and save it
+	if err != nil || token.IsExpired() {
+		// refresh token within a write-lock
+		muCacheLock.Lock()
+
+		// load token again, in case it's been refreshed by another thread
+		token, err = loadToken()
+
+		// check loaded token's error/validity, then refresh/save token
+		if err != nil || token.IsExpired() {
+			// get new token
+			spt, err := refreshToken()
+			if err != nil {
+				return nil, err
+			}
+
+			// use the refreshed token
+			tokenString = spt.OAuthToken()
+		} else {
+			// use locally cached token
+			tokenString = token.OAuthToken()
+		}
+
+		muCacheLock.Unlock()
+	} else {
+		// use locally cached token
+		tokenString = token.OAuthToken()
+	}
+
+	// return acquired token
+	return func() (string, error) {
+		return tokenString, nil
+	}, nil
+}
+
+// Load token from in-mem cache
+func loadToken() (*adal.Token, error) {
+	if adalToken == nil {
+		return nil, fmt.Errorf("Token is nil or Failed to load existing token")
+	}
+
+	return adalToken, nil
+}
+
+// Refresh token for the resource, and save to in-mem cache
+func refreshToken() (*adal.Token, error) {
+	// resource id for Azure SQL Database
+	const sqlAzureResourceID = "https://database.windows.net/"
+
+	// get MSI endpoint to get a token
+	msiEndpoint, err := adal.GetMSIVMEndpoint()
+	if err != nil {
+		return nil, err
+	}
+
+	// get new token for the resource id
+	spt, err := adal.NewServicePrincipalTokenFromMSI(msiEndpoint, sqlAzureResourceID)
+	if err != nil {
+		return nil, err
+	}
+
+	// ensure token is fresh
+	err = spt.EnsureFresh()
+	if err != nil {
+		return nil, err
+	}
```

Comment:
You might want to use
```suggestion
	if err := spt.EnsureFresh(); err != nil {
		return nil, err
	}
```
here.